sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,292 @@
+Since the object hasn't been "ready", there is no reason to add the validation rules required on
+updates since restrictions like field immutability don't logically apply and would restrict the power
+of admission extensions to the point of unusability.  Using the results for creation validation on 
+uninitialized objects and objects transitioning from uninitialized to initialized lets the
+strategy author focus on their API needs.
+
+Admission validation is also needed on resources being initialized.  During initialization, there
+may be some admission validation required, but it isn't normal API update admission.  The update
+admission mutators shouldn't run and since the subject (user.Info) on the admission attributes is
+NOT the original user, normal update admission mutators and validators are likely to make improper
+choices.  When the object transitions from uninitialized to initialized state, the original object
+creator should be artificially injected into the admission attributes and the create admission 
+validators should run.  This allows mechanisms like PSP to function properly during their validation
+phase.  Such injection requires persisting the original user.Info (probably encrypted) into the 
+object.
+
+## How much is needed for beta
+Basically all of it.  After considering the needs of initializers, the admission chain needs to
+be split into mutators and validators.  That has a direct impact on the webhook admission registration
+and callout APIs.  The initializer API handling needs to be updated to work with our existing chain.
+Mutating admission webhooks are needed to support parity in the chain, so we know we'll eventually
+need them and because they will directly interact in the registration API, we need to know how we'll
+build them before we can push a registration API to beta.
+
+
+### Admission interfaces
+
+The current interface in `pkg/admission/interfaces.go` looks like this:
+
+```go
+type Interface interface {
+  Admit(a Attributes) (err error)
+  Handles(operation Operation) bool
+}
+```
+
+We'd like to break this into multiple interfaces, one per phase, called at the appropriate times
+within the request lifecycle.
+
+```go
+type Mutator interface {
+  // ApplyMutations is executed after the request body is deserialized into an
+  // object for create and update requests.
+  // Plugins are free to modify the object as they see fit. If an error is
+  // returned, the request is marked as failed and no changes are persisted.
+  ApplyMutations(a Attributes) error
+}
+
+type Validator interface {
+  // Validate is executed after the appropriate `rest.Storage` handler's
+  // validation occurs. This gives the plugin the opportunity to validate the
+  // object after the handler has validated it, but before it is persisted. If
+  // an error is returned, the request is marked as failed and no changes are
+  // persisted.
+  Validate(a Attributes) error
```

Comment:
so limit range defaulting moves to apply mutation, but limit range enforcement moves to validate.

resource quota enforcement moves to validate only, and we stop having the problem of invalid objects bumping usage.

:100:

Reply:
> resource quota enforcement moves to validate only, and we stop having the problem of invalid objects bumping usage.

Correct.